Implementing transpose and view methods for PackedTensorAccessor

ramvenkat98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60968**

See https://github.com/pytorch/pytorch/issues/60431

The methods are implemented to match the functionality of THCDeviceTensor, which was used for a similar purpose as TensorAccessor/PackedTensorAccessor.

Corner cases are:
1. For the 1d case, transpose returns a PackedTensorAccessor that is an identical copy of the original; this mirrors the behavior of THCDeviceTensor
2. The 1d case for view is not implemented - the reason for this is that in THCDeviceTensor, there would be no 1d case for view - it would be forbidden by static asserts. So this matches that behavior.

Differential Revision: [D29505537](https://our.internmc.facebook.com/intern/diff/D29505537)

Commit:
Implementing transpose and view methods for PackedTensorAccessor

See https://github.com/pytorch/pytorch/issues/60431

The methods are implemented to match the functionality of THCDeviceTensor, which was used for a similar purpose as TensorAccessor/PackedTensorAccessor.

Corner cases are:
1. For the 1d case, transpose returns a PackedTensorAccessor that is an identical copy of the original; this mirrors the behavior of THCDeviceTensor
2. The 1d case for view is not implemented - the reason for this is that in THCDeviceTensor, there would be no 1d case for view - it would be forbidden by static asserts. So this matches that behavior.

Differential Revision: [D29459384](https://our.internmc.facebook.com/intern/diff/D29459384/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29459384/)!

[ghstack-poisoned]

Commit:
Update on "Implementing transpose and view methods for PackedTensorAccessor"


See https://github.com/pytorch/pytorch/issues/60431

The methods are implemented to match the functionality of THCDeviceTensor, which was used for a similar purpose as TensorAccessor/PackedTensorAccessor.

Corner cases are:
1. For the 1d case, transpose returns a PackedTensorAccessor that is an identical copy of the original; this mirrors the behavior of THCDeviceTensor
2. The 1d case for view is not implemented - the reason for this is that in THCDeviceTensor, there would be no 1d case for view - it would be forbidden by static asserts. So this matches that behavior.

Differential Revision: [D29459384](https://our.internmc.facebook.com/intern/diff/D29459384/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29459384/)!

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/TensorAccessor.h:
```diff
@@ -187,6 +196,39 @@ class GenericPackedTensorAccessor : public GenericPackedTensorAccessorBase<T,N,P
+  }
+
+  /// Returns a PackedTensorAccessor that is a view of the `SubDim`-dimensional
+  /// slice of this tensor, starting at `at`
```
This doesn't match the signature of Tensor.view?

ezyang commented on aten/src/ATen/test/packedtensoraccessor_test.cpp:
```diff
@@ -0,0 +1,61 @@
+  for (int i = 0; i < sizes[0]; i++)
+    for (int j = 0; j < sizes[1]; j++)
+      for (int k = 0; k < sizes[2]; k++)
+        ASSERT_EQ(original[i][j][k], transposed[k][j][i]);
```
nit: use braces

ezyang commented:
We're gonna drop view() and this is good to go.

@ramvenkat98 Do you know if it is easy or not to make TensorAccessor also support transpose, but by returning a PackedTensorAccessor? If we gotta copy paste code to make it work, not worth it, we'll just do it as is here.

Commit:
Update on "Implementing transpose and view methods for PackedTensorAccessor"


See https://github.com/pytorch/pytorch/issues/60431

The methods are implemented to match the functionality of THCDeviceTensor, which was used for a similar purpose as TensorAccessor/PackedTensorAccessor.

Corner cases are:
1. For the 1d case, transpose returns a PackedTensorAccessor that is an identical copy of the original; this mirrors the behavior of THCDeviceTensor
2. The 1d case for view is not implemented - the reason for this is that in THCDeviceTensor, there would be no 1d case for view - it would be forbidden by static asserts. So this matches that behavior.

Differential Revision: [D29459384](https://our.internmc.facebook.com/intern/diff/D29459384/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D29459384/)!

[ghstack-poisoned]

ramvenkat98 commented:
Copy-pasting the code would work in letting transpose run (on hosts), but I'm not aware of a cleaner way of doing this.

I guess that in general, as more functions are added to this file, there might be a decent amount of code duplication to deal with (since the implementations have to be written for both the base case and the recursive case, for both TensorAccessor and PackedTensorAccessor).

As a random thought, one way of getting around this is might be to make TensorAccessor a base class and PackedTensorAccessor inherit from it but with PackedTensorAccessor having two extra data members - the size and stride array. In this case, it would be possible to implement most methods directly in TensorAccessor (like transpose - which will always return a PackedTensorAccessor - and I think even the existing [] operators - which always return a TensorAccessor - would not need to be duplicated).

There might be some extra difficulties to this that I'm missing (like the host/device stuff), and it adds one extra level of inheritance since now there's PackedTensorAccessor <- TensorAccessor <- TensorAccessor base but it can help get rid of PackedTensorAccessorBase and perhaps avoid some duplication...

ezyang commented:
Approved. For the hypothetical other improvements, another way to do it without duplication and without inheritance is to set up a traits class specifying what operations PackedTensorAccessor and TensorAccessor support, and then implement transpose generically over the trait.

ramvenkat98 commented:
@ramvenkat98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29459384).

ramvenkat98 commented:
@ramvenkat98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29459384).

driazati commented:
@driazati has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29503741).

ramvenkat98 commented:
@ramvenkat98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29505537).

ramvenkat98 commented:
Closed: see https://github.com/pytorch/pytorch/pull/61114 instead
